configure

```diff
@@ -3642,28 +3649,41 @@ echo_n "Checking whether to have a verbose build... "
 echo "$_verbose_build"
 add_to_config_mk_if_yes "$_verbose_build" 'VERBOSE_BUILD = 1'
 
-
-#
-# If a specific optimization level was requested, enable optimizations
-#
-if test -n "$_optimization_level" ; then
-	# Ports will specify an optimization level and expect that to be enabled
-	if test "$_optimizations" != no ; then
-		_optimizations=yes
-	fi
-else
-	_optimization_level=$_default_optimization_level
-fi
-
 #
 # Check whether to enable optimizations
 #
+echo_n "Checking best optimization level... "
+echo "int main() { return 0; }" > $TMPC
 if test "$_optimizations" = yes ; then
-	# Enable optimizations. This also
-	# makes it possible to use -Wuninitialized, so let's do that.
+	if test -z "$_optimization_level" ; then
+		# For pretty much all systems, -Os is the best level if it is supported
+		# by the compiler, since old devices are memory-constrained and (as a
+		# default optimization strategy) new devices are generally fastest when
+		# cache misses are minimized by keeping the code small as well.
+		cc_check_no_clean -Os && _optimization_level=-Os || _optimization_level=-O2
+		echo_n -- "$_optimization_level"
+
+		if cc_check_no_clean $_optimization_level -ftree-vectorize; then
+			echo_n " with auto-vectorization"
+			append_var CXXFLAGS "-ftree-vectorize"
+		fi
+	else
+		echo_n -- "$_optimization_level (forced)"
+	fi
+	echo ""
 	append_var CXXFLAGS "$_optimization_level"
+
+	# Enabling optimizations also makes it possible to use -Wuninitialized, so
+	# let's do that.
 	append_var CXXFLAGS "-Wuninitialized"
+elif test "$_optimizations" != no && cc_check_no_clean -Og; then
+	echo "-Og"
```

Comment:
Are you sure about setting `-Og` by default ? It's good for us for building faster debug builds, but won't linux packages maintainers hate it? For example the Gentoo people like setting very specific compiler flags via environment variables.

Reply:
I’m not very confident. I’ve waffled back and forth on the best non-release optimisation strategy, and on whether or not Buildbot should just always generate fully optimised builds since users today use these builds as rolling releases. It seemed like *something* needed to be done differently since `-O0` isn’t good enough for Munt, and `-Og` seemed like it could be a decent compromise between runtime speed, compilation speed, and avoiding symbols optimised away. As a result of my doubt, I added the user override so it is possible to force the strategy with e.g. `--enable-optimizations=-O0` if needed.

Reply:
Sorry, my comment was not clear enough. `-Og` seems like a good compromise to me for debug builds.
What I'm not sure about is if configure should set any optimization flags by default. Linux packaging tools usually set the optimisation flags in the environment themselves. If we override that with `-Og`, people will expect to have release builds but instead get debug-optimized builds.

Reply:
I've looked at the ScummVM Gentoo ebuild and it doesn't seem to use the system's `C[XX]FLAGS` specified in `/etc/make.conf` right now. Usually, it is expected that ebuilds honor the environment variables to an extent that makes sense for the given package. When upstream doesn't use `autoconf` and "unnecessarily" enforces or hardcodes certain compiler or linker flags Gentoo usually strips these away (either on-the-fly as part of the ebuild or via a separate patch file). If we don't have this at the moment, we should allow overriding `C[XX]FLAGS` via the commandline. As long as that is possible I see no problem defaulting to `-Og`.

Reply:
Let me clarify: I see no problem defaulting to `-Og` for *debug builds*, just as bgK says.

Reply:
OK, I understand now, thank you for clarifying. So is the best approach then to not do this at all without an opt-in, not do this if `CXXFLAGS` is set, try to test for an `-O` flag in `CXXFLAGS` and not do it if one is found, something else?

Reply:
I'm very unsure what's the best option. We seem to enable debug builds by default. Perhaps when `--disable-debug` is passed to configure then `-Og` should not be set. That way it's possible not to have optimization flags without editing configure.

Reply:
Oh, but it's also possible to pass `--disable-optimizations` to disable `-Og`. Then I guess it's good.

Reply:
Sounds good.